src/Perpetuum/Services/EventServices/EventProcessors/NpcSpawnEventHandlers/NpcReinforcementSpawner.cs

```diff
@@ -0,0 +1,102 @@
+    {
+        protected override TimeSpan SPAWN_DELAY { get { return TimeSpan.FromSeconds(10); } }
+        protected override TimeSpan SPAWN_LIFETIME { get { return TimeSpan.FromMinutes(15); } }
+        protected override int MAX_SPAWN_DIST { get { return 10; } }
+
+        private readonly IDictionary<NpcBossInfo, INpcReinforcements> _reinforcementsByNpc = new Dictionary<NpcBossInfo, INpcReinforcements>();
+        public NpcReinforcementSpawner(IZone zone, INpcReinforcementsRepository reinforcementsRepo) : base(zone, reinforcementsRepo) { }
+
+        protected override void CheckReinforcements(NpcReinforcementsMessage msg)
+        {
+            var info = msg.Npc.BossInfo;
+            if (!_reinforcementsByNpc.ContainsKey(info))
+            {
+                var reinforcements = _npcReinforcementsRepo.CreateNpcBossAddSpawn(info, msg.ZoneId);
+                _reinforcementsByNpc.Add(info, reinforcements);
+            }
+        }
+
+        protected override bool CheckState(NpcReinforcementsMessage msg)
+        {
+            if (msg.Npc.BossInfo.IsDead)
+            {
+                CleanupAllReinforcements(msg);
+                return true;
+            }
+            return false;
+        }
+
+        protected override void CleanupAllReinforcements(NpcReinforcementsMessage msg)
+        {
+            var info = msg.Npc.BossInfo;
+            if (_reinforcementsByNpc.ContainsKey(info))
+            {
+                var activeWaves = _reinforcementsByNpc[info].GetAllActiveWaves();
+                foreach (var wave in activeWaves)
+                {
+                    ExpireWave(wave);
+                }
+                _reinforcementsByNpc.Remove(info);
+            }
+        }
+
+        protected override Position FindSpawnPosition(NpcReinforcementsMessage msg, int maxRange)
+        {
+            var finder = new RandomWalkableAroundPositionFinder(_zone, msg.Npc.CurrentPosition, maxRange);
+            if (finder.Find(out Position result))
+            {
+                return result;
+            }
+            return msg.Npc.CurrentPosition;
+        }
+
+        protected override INpcReinforcementWave GetNextWave(NpcReinforcementsMessage msg)
+        {
+            var npc = msg.Npc;
+            var percent = 1.0 - npc.ArmorPercentage;
+            return _reinforcementsByNpc[npc.BossInfo].GetNextPresence(percent);
+        }
+
+        protected override void OnSpawning(Presence pres, NpcReinforcementsMessage msg)
+        {
+            foreach (var npc in pres.Flocks.GetMembers())
+            {
+                foreach (var threat in msg.Npc.ThreatManager.Hostiles)
+                {
+                    npc.AddDirectThreat(threat.unit, threat.Threat + FastRandom.NextDouble(5, 10));
+                }
+            }
+        }
+
+        protected override List<INpcReinforcements> GetActiveReinforcments(Presence presence)
```

Comment:
Return `ReadOnlyCollection` unless you explicitly want the caller to be able to modify the list.

Also, unless you want this immediately enumerating, might be worth retuning an `IEnumerable` in this case depends on the lifetime of the parent.